python/tvm/relax/base_py_module.py

```diff
@@ -0,0 +1,367 @@
+    ) -> Union[NDArray, List[NDArray]]:
+        """Convert PyTorch tensors to TVM NDArrays using DLPack."""
+        if isinstance(tensors, (list, tuple)):
+            return [self._convert_single_pytorch_to_tvm(t) for t in tensors]
+        return self._convert_single_pytorch_to_tvm(tensors)
+
+    def _convert_single_pytorch_to_tvm(self, tensor: Any) -> NDArray:
+        """Convert a single PyTorch tensor to TVM NDArray with robust fallbacks."""
+        if isinstance(tensor, NDArray):
+            return tensor
+        if isinstance(tensor, torch.Tensor):
+            # 1. Try modern `torch.to_dlpack` (preferred for PyTorch >= 1.7)
+            try:
+                dlpack = torch.to_dlpack(tensor)
+                return tvm.nd.from_dlpack(dlpack)
+            except (AttributeError, ValueError):
+                pass  # Fall through to the next method
+            # 2. Try legacy `torch.utils.dlpack.to_dlpack`
+            if to_dlpack_legacy:
+                try:
+                    dlpack = to_dlpack_legacy(tensor)
+                    return tvm.nd.from_dlpack(dlpack)
+                except (AttributeError, ValueError) as error_legacy:
+                    print(
+                        f"Warning: Legacy DLPack conversion failed ({error_legacy}), "
+                        f"using numpy fallback."
+                    )
+            # 3. If all DLPack methods fail, use numpy fallback
+            numpy_array = tensor.detach().cpu().numpy()
+            return tvm.nd.array(numpy_array, device=self.device)
+
+        # For other types (like scalars, lists), convert to numpy first
+        try:
+            numpy_array = np.array(tensor, dtype=np.float32)
+            return tvm.nd.array(numpy_array, device=self.device)
+        except (TypeError, ValueError) as error:
+            raise TypeError(
+                f"Unsupported type for conversion to TVM NDArray: {type(tensor)}"
+            ) from error
+
+    def _convert_tvm_to_pytorch(
+        self, tvm_arrays: Union[Any, List[Any]]
+    ) -> Union[torch.Tensor, List[torch.Tensor]]:
+        """Convert TVM NDArrays to PyTorch tensors using DLPack."""
+        if isinstance(tvm_arrays, (list, tuple)):
+            return [self._convert_single_tvm_to_pytorch(arr) for arr in tvm_arrays]
+        return self._convert_single_tvm_to_pytorch(tvm_arrays)
+
+    def _convert_single_tvm_to_pytorch(self, tvm_array: Any) -> torch.Tensor:
+        """Convert a single TVM NDArray to PyTorch tensor using DLPack."""
+        if isinstance(tvm_array, torch.Tensor):
+            return tvm_array
+        if not isinstance(tvm_array, NDArray):
+            return torch.tensor(tvm_array)
+        try:
+            dlpack = tvm_array.to_dlpack()
+            return torch.from_dlpack(dlpack)
+        except (tvm.TVMError, RuntimeError) as error:
+            print(f"Warning: DLPack conversion from TVM failed ({error}), using numpy fallback")
+            numpy_array = tvm_array.asnumpy()
```

Comment:
```suggestion
            numpy_array = tvm_array.numpy()
```
`asnumpy` is formally removed by the latest commits. Let's use `.numpy()`.  Likely we also need to rebase this PR onto the latest main branch.